src/Transports.AspNetCore/WebSockets/SubscriptionList.cs

```diff
@@ -0,0 +1,228 @@
+/// <see cref="Dispose"/> will throw an <see cref="OperationCanceledException"/>.
+/// <br/><br/>
+/// It is recommended to call <see cref="Dispose"/> immediately after signalling
+/// the cancellation token.
+/// </summary>
+public sealed class SubscriptionList : IDisposable
+{
+    private readonly CancellationToken _cancellationToken;
+    private readonly Dictionary<string, IDisposable> _subscriptions = new();
+    private readonly object _lock = new();
+
+    /// <summary>
+    /// Initializes a new instance.
+    /// </summary>
+    public SubscriptionList(CancellationToken cancellationToken)
+    {
+        _cancellationToken = cancellationToken;
+    }
+
+    /// <summary>
+    /// Disposes of all active subscriptions.
+    /// </summary>
+    public void Dispose()
+    {
+        List<IDisposable> subscriptionsToDispose;
+        lock (_lock)
+        {
+            subscriptionsToDispose = _subscriptions.Values.ToList();
+            _subscriptions.Clear();
+        }
+        foreach (var disposer in subscriptionsToDispose)
+        {
+            disposer.Dispose();
+        }
+    }
+
+    /// <summary>
+    /// Adds a subscription to the internal list if none is already in the list with the same id.
+    /// </summary>
+    /// <exception cref="ArgumentNullException"/>
+    /// <exception cref="OperationCanceledException"/>
+    public bool TryAdd(string id, IDisposable subscription)
+    {
+        if (id == null)
+            throw new ArgumentNullException(nameof(id));
+        if (subscription == null)
+            throw new ArgumentNullException(nameof(subscription));
+
+        lock (_lock)
+        {
+            _cancellationToken.ThrowIfCancellationRequested();
+            return _subscriptions.TryAdd(id, subscription);
+        }
+    }
+
+    /// <summary>
+    /// Adds a subscription to the internal list, overwriting an existing registration if any.
+    /// When overwriting an existing registration, the old registration is disposed.
+    /// </summary>
+    /// <exception cref="ArgumentNullException"/>
+    /// <exception cref="OperationCanceledException"/>
+    public IDisposable this[string id]
+    {
+        set
+        {
+            if (id == null)
+                throw new ArgumentNullException(nameof(id));
+            if (value == null)
+                throw new ArgumentNullException(nameof(value));
+
+            IDisposable? oldDisposable = null;
+            try
+            {
+                lock (_lock)
+                {
+                    _cancellationToken.ThrowIfCancellationRequested();
+                    _subscriptions.TryGetValue(id, out oldDisposable);
+                    _subscriptions[id] = value;
+                }
+            }
+            finally
+            {
+                oldDisposable?.Dispose();
+            }
+        }
+    }
+
+    /// <summary>
+    /// Validates that the specified subscription is still active.
```

Comment:
```suggestion
    /// Validates that the specified subscription is still in the list.
```

I would avoid term _active_ .